posthog/api/test/test_plugin.py

```diff
@@ -463,40 +460,43 @@ def test_create_plugin_version_range_gt_next_major_ignore_on_cloud(self, mock_ge
             )
             self.assertEqual(response.status_code, 201)
 
-    def test_create_plugin_source(self, mock_get, mock_reload):
+    def test_update_plugin_source(self, mock_get, mock_reload):
         self.assertEqual(mock_reload.call_count, 0)
         response = self.client.post(
-            "/api/organizations/@current/plugins/",
-            {"plugin_type": "source", "name": "myplugin", "source": "const processEvent = e => e",},
-        )
-        self.assertEqual(response.status_code, 201)
-        self.assertEqual(
-            response.json(),
-            {
-                "id": response.json()["id"],
-                "plugin_type": "source",
-                "name": "myplugin",
-                "description": None,
-                "url": None,
-                "config_schema": {},
-                "tag": None,
-                "source": "const processEvent = e => e",
-                "latest_tag": None,
-                "is_global": False,
-                "organization_id": response.json()["organization_id"],
-                "organization_name": self.CONFIG_ORGANIZATION_NAME,
-                "capabilities": {},
-                "metrics": {},
-                "public_jobs": {},
-            },
+            "/api/organizations/@current/plugins/", {"plugin_type": "source", "name": "myplugin_original",},
         )
-        self.assertEqual(Plugin.objects.count(), 1)
+        plugin_id = response.json()["id"]
         self.assertEqual(mock_reload.call_count, 1)
 
-        self.client.delete("/api/organizations/@current/plugins/{}".format(response.json()["id"]))
-        self.assertEqual(Plugin.objects.count(), 0)
+        response = self.client.get(f"/api/organizations/@current/plugins/{plugin_id}/source")
+        self.assertEqual(response.json(), {})
+        self.assertEqual(Plugin.objects.get(pk=plugin_id).name, "myplugin_original")
+
+        response = self.client.patch(
+            f"/api/organizations/@current/plugins/{plugin_id}/update_source",
+            data=json.dumps({"index.ts": "'hello world'", "plugin.json": '{"name":"my plugin"}'}),
+            content_type="application/json",
+        )
+        self.assertEqual(response.json(), {"index.ts": "'hello world'", "plugin.json": '{"name":"my plugin"}'})
+        self.assertEqual(Plugin.objects.get(pk=plugin_id).name, "my plugin")
         self.assertEqual(mock_reload.call_count, 2)
 
+        response = self.client.patch(
+            f"/api/organizations/@current/plugins/{plugin_id}/update_source",
+            data=json.dumps({"index.ts": "'hello again'"}),
+            content_type="application/json",
+        )
+        self.assertEqual(response.json(), {"index.ts": "'hello again'", "plugin.json": '{"name":"my plugin"}'})
+        self.assertEqual(mock_reload.call_count, 3)
+
+        response = self.client.patch(
```

Comment:
Q: Why do we have a test that tests N different scenarios? This makes for a hard to understand and debug test suite. Could we give each scenario a name and test separately?

Reply:
I decided against that initially to avoid needless scaffolding. The test essentially sets up a story, and then tests based on that story, in 5 steps. I can change maximally optimise this to be three tests with 3, 4 and 4 steps respectively. This seemed wasteful, so I opted to have one 5-step story test the entire flow.

I added helpful hints to make it easier to follow, but can break it up if you think this makes sense.